community/kernel/src/main/java/org/neo4j/kernel/impl/index/IndexDefineCommand.java

```diff
@@ -129,10 +129,10 @@ private int getOrAssignId( Map<String,Integer> stringToId, PrimitiveIntObjectMap
         }
 
         int nextIdInt = nextId.incrementAndGet();
-        if ( nextIdInt > HIGHEST_POSSIBLE_ID ) // >= since the actual value -1 is reserved for all-ones
+        if ( nextIdInt > HIGHEST_POSSIBLE_ID || stringToId.size() > HIGHEST_POSSIBLE_ID )
         {
             throw new IllegalStateException( format(
-                    "Modifying more than %d indexes in a single transaction is not supported",
+                    "Modifying more than %d indexes or keys in a single transaction is not supported",
```

Comment:
It looks like that result message will show number that is bigger then actual maximum because of +1.